pkg/config/remote_config.go

```diff
@@ -0,0 +1,142 @@
+package config
+
+import (
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"os"
+	"strings"
+
+	"github.com/prometheus/common/config"
+)
+
+// supported remote config provider schemes
+const (
+	HTTP  = "http"
+	HTTPS = "https"
+	// TODO: add s3, gcs, blob, and git providers backed by go-fsimpl
+)
+
+// RemoteOpts struct contains agent remote config options
+type RemoteOpts struct {
+	ExpandEnv bool
+
+	URL              *url.URL
+	HTTPClientConfig *config.HTTPClientConfig
+}
+
+// RemoteProvider ...
+type RemoteProvider interface {
+	Retrieve() (*Config, error)
+}
+
+// RemoteConfig ...
+type RemoteConfig struct {
+	RemoteProvider
+}
+
+// NewRemoteConfig ...
+func NewRemoteConfig(rawURL string, opts *RemoteOpts) (*RemoteConfig, error) {
+	u, err := url.Parse(rawURL)
+	if err != nil {
+		return nil, err
+	}
+	if opts == nil {
+		// Default provider opts
+		opts = &RemoteOpts{
+			ExpandEnv: false,
+		}
+	}
+	opts.URL = u
+
+	switch u.Scheme {
+	case "":
+		// if no scheme, assume local file path, return nil and let caller handle.
+		return nil, nil
+	case HTTP, HTTPS:
+		return &RemoteConfig{
+			RemoteProvider: newHTTPProvider(opts),
+		}, nil
+	}
+	return nil, fmt.Errorf("remote config scheme not supported: %s", u.Scheme)
+}
+
+// Remote Config Providers
+// httpP - http/https provider
+type httpP struct {
+	myURL            *url.URL
+	expandEnv        bool
+	httpClientConfig *config.HTTPClientConfig
+}
+
+func newHTTPProvider(opts *RemoteOpts) *httpP {
+	return &httpP{
+		myURL:            opts.URL,
+		expandEnv:        opts.ExpandEnv,
+		httpClientConfig: opts.HTTPClientConfig,
+	}
+}
+
+// Retrieve implements RemoteProvider and fetches the config
+// TODO: token auth, oauth2, etc.
+func (p httpP) Retrieve() (*Config, error) {
+	var (
+		bb  []byte
+		err error
+
+		request   *http.Request
+		response  *http.Response
+		basicAuth *config.BasicAuth
+
+		result = &Config{}
+		client = &http.Client{}
+	)
+
+	if p.httpClientConfig != nil {
```

Comment:
nit: Use `config.NewClientFromConfig` here and call `SetDirectory` from the caller instead of in the handler

Reply:
@rfratto any reason we want to use the prometheus `http.Client` vs using Go's `http.DefaultClient`?  I didn't look into what all the differences were but assumed the prometheus client was likely optimized for pushing metrics

Reply:
I think if we're going to use Prometheus' HTTP client config, then we should reuse their existing code that generates a client from it :) It's not very Prometheus-specific, anything in prometheus/common is pretty broadly usable.

Reply:
> If we suspect that we might also support remote_urls for #868, then it might be worthwhile having RemoteConfig implement fs.FS sooner rather than later. Otherwise, if we don't think we're going to expand the remote URL concept any further than we probably don't need a broader interface yet.

This makes sense to me. I intentionally kept this simple / extendable so it shouldn't be a problem to support conf.d style configs if we decide to in the future. Instead of having future requirements creep into the scope of this feature, I think we should implement local conf.d configuration support first, and then add it for remote as a separate piece of work.